bitswap/testinstance/testinstance.go

```diff
@@ -49,12 +49,28 @@
 
 // Next generates a new instance of bitswap + dependencies
 func (g *InstanceGenerator) Next() Instance {
+	return g.NextWithExtraOptions(nil)
+}
+
+// NextWithExtraOptions is like [Next] but it will callback with a fake identity and append extra options.
+// If extraOpts is nil, it will ignore it.
+func (g *InstanceGenerator) NextWithExtraOptions(extraOpts func(p tnet.Identity) ([]bsnet.NetOpt, []bitswap.Option)) Instance {
 	g.seq++
 	p, err := p2ptestutil.RandTestBogusIdentity()
 	if err != nil {
-		panic("FIXME") // TODO change signature
+		panic(err.Error()) // TODO change signature
 	}
-	return NewInstance(g.ctx, g.net, p, g.netOptions, g.bsOptions)
+
+	var extraNet []bsnet.NetOpt
+	var extraBitswap []bitswap.Option
+	if extraOpts != nil {
+		extraNet, extraBitswap = extraOpts(p)
+	}
+
+	return NewInstance(g.ctx, g.net, p,
+		append(g.netOptions[:len(g.netOptions):len(g.netOptions)], extraNet...),
+		append(g.bsOptions[:len(g.bsOptions):len(g.bsOptions)], extraBitswap...),
```

Comment:
What does `g.netOptions[:len(g.netOptions):len(g.netOptions)` do? Bit confused with the double semicolons 😅

Reply:
This sets leave the length unchanged and sets the capacity to the length, this force `append` to make a copy, else the array is still shared and this can cause memory corruptions when improperly done.

Reply:
Intriguing, thanks.